### PR TITLE
Add settings for custom base url

### PR DESCRIPTION
**Pull request by DGdev91**
Making the openai base url and embedding dimension configurable, these are useful to integrate AutoGPT with other models, like LLaMA

### Background
This makes AutoGPT capable of connecting to custom openai-like APIs like [[keldenl](https://github.com/keldenl)](https://github.com/keldenl/gpt-llama.cpp), and use other models, like LLaMA and derivates.

see also #25 #567  #2158

### Changes
Added OPENAI_API_BASE_URL ~~and EMBED_DIM~~ to .env_template and loaded them in config.py, making sure OPENAI_API_BASE_URL would be ignored if USE_AZURE is True.

~~Also, modified the files in autogpt/memory to use the value in EMBED_DIM instead of 1536 (wich is still the default)~~

UPDATE: embed_dim settings isn't needed anymore

### Documentation
I added an explanation of what those new configurations do in the .env_template file, following the comments on other configurations

### Test Plan
Tested it by using gpt-llama.cpp on my machine, and setting OPENAI_API_BASE_URL to the API url in my .env file.
~~I used Vicuna 13B, so i also set EMBED_DIM to 5120~~
For this test, i also set  OPENAI_API_KEY to the model's path (it's an "hack" made by gpt-llama.cpp to get the model's path)

### PR Quality Checklist
- * [x] My pull request is atomic and focuses on a single change.
- * [x] I have thoroughly tested my changes with multiple different prompts.
- * [x] I have considered potential risks and mitigations for my changes.
- * [x] I have documented my changes clearly and comprehensively.
- * [x] I have not snuck in any "extra" small tweaks changes

### CODE REVIEW

**Review by keldenl — APPROVED** [review]
+1, this is good for flexibility – LGTM

**Comment by MarkSchmidty** [comment]
LGTM

**Comment by Sorann753** [comment]
A name like `LLM_API_BASE_URL` instead of `OPENAI_API_BASE_URL` might be more fitting since it allows us to not always use OpenAI's API

**Comment by MarkSchmidty** [comment]
It's still using the OpenAI API, just not their endpoint, even if the model behind it isn't an OpenAI model.

**Comment by cryptocake** [comment]
LGTM 👍

**Review by khalek1414 — APPROVED** [review]

**Review by keldenl — DISMISSED** [review]
Reapproving it because this still LGTM and is awesome

File: .env.template
Comment (ntindle) [inline .env.template]: Should this be openai base url or LLM base url?

File: .env.template
Comment (ntindle) [inline .env.template]: This seems fraught for mistakes is there anywhere these can be pulled from in the codebase?

File: autogpt/config/config.py
Comment (ntindle) [inline autogpt/config/config.py]: I really don’t love the magic number-ness of this here

File: autogpt/memory/local.py
Comment (ntindle) [inline autogpt/memory/local.py]: Config loading process was chnaged. Update from master and use the new method CC @collijk

File: autogpt/memory/redismem.py
Comment (ntindle) [inline autogpt/memory/redismem.py]: Config loading same as above @collijk

**Review by ntindle — CHANGES_REQUESTED** [review]

File: .env.template
Comment (keldenl) [reply .env.template]: since this is specifically to redirect openai-like traffic to a local or different host, i would say OPEN_API_BASE_URL is appropriate here!

Comment (DGdev91) [reply .env.template]: Yeah, i see "OPENAI_API_BASE_URL" as more appropriate because it still uses OpenAI's API

File: .env.template
Comment (DGdev91) [reply .env.template]: As you can see from the changes, EMBED_DIM was hardcoded as 1536 in every case it was used before.
I'm leaving 1536 as default and letting the user use whatever he needs.
Those example have been found by Kelndenl (https://github.com/keldenl/gpt-llama.cpp/issues/2#issuecomment-1514353829), they are from here
https://huggingface.co/shalomma/llama-7b-embeddings#quantitative-analysis

Anyway, it's just an example. If no one touches that value, it remains 1536 just as before

File: autogpt/config/config.py
Comment (DGdev91) [reply autogpt/config/config.py]: 1536 as you can see was hardcoded everywhere. i just left it as default value.

File: autogpt/memory/local.py
Comment (DGdev91) [reply autogpt/memory/local.py]: I merged the last changes and looked at that but i don't know what are you referring to. Later in the code the configuration was passed directly in the class functions, but i still need this code for the global variable `EMBED_DIM = CFG.embed_dim`
Many classes around the code still use cfg = Config(), is it that wrong?

File: autogpt/memory/redismem.py
Comment (DGdev91) [reply autogpt/memory/redismem.py]: Same as before, if you saying "use the cfg passed to the class functions", it's not pussible here, i need it in a global variable
```
CFG = Config()
SCHEMA = [
    TextField("data"),
    VectorField(
        "embedding",
        "HNSW",
        {"TYPE": "FLOAT32", "DIM": 1536, "DISTANCE_METRIC": "COSINE"},
        {"TYPE": "FLOAT32", "DIM": CFG.embed_dim, "DISTANCE_METRIC": "COSINE"},
    ),
]
```

If it's not that, i don't know what are you referring to

File: autogpt/memory/local.py
Comment (cryptocake) [reply autogpt/memory/local.py]: @ntindle `local.py` has `EMBED_DIM = 1536` hardcoded outside of any class scope at top of the file. It is only being used in another non-class function:
```python
def create_default_embeddings():
    return np.zeros((0, EMBED_DIM)).astype(np.float32)
```

**Review by cryptocake — APPROVED** [review]
Proposed changes look good. Embedding dimensions should not be hardcoded and need to be manipulated if Auto-GPT ever wishes to support local LLM's.

**Comment by jvgriethuijsen** [comment]
Thanks so much for building this @DGdev91 and delivering the required documentation. Really awesome job!

For the ones struggling to implement this, it took me a while finding the right model for the job. Eventually I got it to work with ggml-vicuna-13b-1.1-q4_2.bin (from huggingface).

my .env.: 
- OPENAI_API_BASE_URL=http://localhost:443/v1
- EMBED_DIM=5120
- OPENAI_API_KEY=M:\AI\llama.cpp\models\ggml-vicuna-13b-1.1-q4_2.bin

I do have to say, it's incredibally slow on my machine. While I have a decent processor and 32G of ram (and a geforce RTX 3070ti) and am running the model from a fast SSD, it will not utilize my full machine. It will actually timeout (600 seconds) every request unless I put the TIMEOUT_SECS = 6000 in the api_requestor.py file of autoGPT. The 7B models were a bit faster, but weren't able to respond in the way that allows autoGPT to actually work. I'm thinking of trying to get it to work with my videocard, since it is the most high end part of my pc, but am not quite sure yet where to start. Will let you know if I make it :)

File: autogpt/memory/redismem.py
Comment (ntindle) [inline autogpt/memory/redismem.py]: Can this be moved into the class and use the passed in cfg.embed_dim?

File: autogpt/memory/local.py
Comment (ntindle) [reply autogpt/memory/local.py]: can create_default_embeddings be moved inside the class so the cfg is accessible?

File: autogpt/memory/redismem.py
Comment (ntindle) [reply autogpt/memory/redismem.py]: can that global variable be moved to not global?

File: autogpt/memory/local.py
Comment (ntindle) [reply autogpt/memory/local.py]: but tldr we are trying to get rid of the cfg=Config() at certain levels so we can initialize multiple instances of certain things, like memory

**Review by ntindle — CHANGES_REQUESTED** [review]

**Comment by ntindle** [comment]
Sorry for such a long back and forth. Want to make sure this is abstracted just enough that we don't have to redo it and break it all

**Comment by ntindle** [comment]
I haven't tested it AT ALL but there's some context for what I'm referring to by the requested changes in the branch: base-url-and-embeddings

**Comment by jmtatsch** [comment]
> I'm thinking of trying to get it to work with my videocard, since it is the most high end part of my pc, but am not quite sure yet where to start. Will let you know if I make it :)

Compile your local api provider with CUBLAS eg for llama-cpp-python 
`LLAMA_CUBLAS=1 pip install llama-cpp-python[server]`

File: autogpt/memory/redismem.py
Comment (DGdev91) [reply autogpt/memory/redismem.py]: Definitley possible, i just kept it as it was before to touch less code as possible. but if you prefer to get rid of cfg=Config() i'll work on it

**Comment by jvgriethuijsen** [comment]
> > > I'm thinking of trying to get it to work with my videocard, since it is the most high end part of my pc, but am not quite sure yet where to start. Will let you know if I make it :)
> > 
> > 
> > Compile your local api provider with CUBLAS eg for llama-cpp-python `LLAMA_CUBLAS=1 pip install llama-cpp-python[server]`
> 
> I guess he's using keldenl's gpt-llama.cpp
> 
> but it can be applied also there. it uses ggerganov's llama.cpp
> 
> git clone https://github.com/ggerganov/llama.cpp.git cd llama.cpp.git make LLAMA_CUBLAS=1
> 
> Of course you need to have CUDA sdk installed for doing that

Thanks both! I'm indeed using keldenl's gpt-llama.cpp currently, but will try your suggestion! I hope I can just direct the OPENAI_API_BASE_URL to the llama-cpp-python[server].
(PS: today autoGPT actually reached my 6000 second request timeout as well, so need to find a better solution xD)

**Comment by DGdev91** [comment]
> Thanks both! I'm indeed using keldenl's gpt-llama.cpp currently, but will try your suggestion! I hope I can just direct the OPENAI_API_BASE_URL to the llama-cpp-python[server]. (PS: today autoGPT actually reached my 6000 second request timeout as well, so need to find a better solution xD)

don't get confused, keldenl's project uses the standard llama.cpp binary, wich is written in cpp. llama-cpp-python is a different proect (python bindings for llama.cpp)

I suggest you to run llama.ccp alone to verify it's compiled correctly and it's actually using the cpu. if it's using cuBLAS, you should see "blas=1" after it loaded the model.
If you are using the same projects you were using the first time, you most likely need to run "make clean" before building it with cuBLAS support.

File: autogpt/memory/redismem.py
Comment (DGdev91) [reply autogpt/memory/redismem.py]: Done

File: autogpt/memory/redismem.py
Comment (DGdev91) [reply autogpt/memory/redismem.py]: Done

File: autogpt/memory/local.py
Comment (DGdev91) [reply autogpt/memory/local.py]: Ok, done.
In the last commit i had to keep the hardcoded default value of 1536 because CacheContent uses dataclasses and needs a default value for his definition, however that value is't actually used, because that class is always initialized by passing embedded_dim from config now.

I tried with both GPT3.5 and Vicuna, seems ok

**Comment by DGdev91** [comment]
> This PR conflicts with #3222 and is not atomic. Please fix that so we can review it.

Why are you saying that? the hardcoded embedding dimension using in memory-related classes and those settings wich he's adding are different things. there are no conflicts.
We also modified different files, only .env.template  and config.py are in common

File: autogpt/config/config.py
Comment (Pwuts) [inline autogpt/config/config.py:86]: Please submit this separately, as it is unrelated to the rest of the PR

**Review by Pwuts — CHANGES_REQUESTED** [review]

**Comment by Pwuts** [comment]
> > This PR conflicts with #3222 and is not atomic. Please fix that so we can review it.
> 
> Why are you saying that? the hardcoded embedding dimension using in memory-related classes and those settings wich he's adding are different things. there are no conflicts. We also modified different files, only .env.template and config.py are in common

Sorry, I could have been more clear, see the comment above. Unrelated changes should not be submitted together, since that makes it harder to review and pick PRs that we want to process.

**Review by ntindle — APPROVED** [review]

**Comment by ntindle** [comment]
Can I get a test to coverage this

**Comment by DGdev91** [comment]
> Sorry, I could have been more clear, see the comment above. Unrelated changes should not be submitted together, since that makes it harder to review and pick PRs that we want to process.

Those change are all about new configurations wich aim to make possible the use of different LLMs, as long as they use an API compliant to OpenAI's, so it made sense to me to put them together.

But if you prefer, i can keep this PR only for EMBED_DIM and put OPENAI_API_BASE_URL in another one

**Comment by DGdev91** [comment]
> Can I get a test to coverage this

If you are referring to the automatic tests wich are making codeconv/patch to fail, most of the uncovered lines are from pinecone and redis integrations, wich don't have tests at all. they never had one even before my changes.

The milvus test resulted uncovered  because i forced the type for the cfg argument in init_collection. made a commit wich should fix that (at least, i hope).

there's also an uncovered line in config.py because (of course) we never set openai.api_base unless we have OPENAI_API_BASE_URL set. Should i add a definition in conftest.py just fot that?

**Comment by DGdev91** [comment]
My last attempt on fixing the milvus_memory_test.py test didn't really had the desired result and CodeCov still marks it as uncovered (the test itself still runs fine). But i'm sure it actually is covered, that code is in the __init__ method and the class is initilized in both milvus_memory_tests.py files.
I guess it's because of that MockConfig object in tests/milvus_memory_test.py
Isn't it better to just initialize a new Config() class like the test under the integration folder already does?

**Comment by p-i-** [comment]
This is a mass message from the AutoGPT core team.
Our apologies for the ongoing delay in processing PRs.
This is because we are re-architecting the AutoGPT core!

For more details (and for infor on joining our Discord), please refer to:
  https://github.com/Significant-Gravitas/Auto-GPT/wiki/Architecting

**Comment by MarkSchmidty** [comment]
> This is a mass message from the AutoGPT core team. Our apologies for the ongoing delay in processing PRs. This is because we are re-architecting the AutoGPT core!
> 
> For more details (and for infor on joining our Discord), please refer to: [Significant-Gravitas/Auto-GPT/wiki/Architecting](https://github.com/Significant-Gravitas/Auto-GPT/wiki/Architecting)

Please merge this PR to master before re-integration. CC @Significant-Gravitas, @Torantulino   
  
Lots of work has gone into it, it's working great in a fork, and it is a very significant upgrade to the base Auto-GPT; providing functionality which is important to the "core" of Auto-GPT going forward.

**Comment by anonhostpi** [comment]
I don't think you quite understand why they aren't merging. The reason for it is the re-arch is going to invalidate all current PRs, because it is going to introduce massive breaking changes to how AutoGPT works. Also not a good idea to beg for merge IMO.

**Comment by DGdev91** [comment]
> I don't think you quite understand why they aren't merging. The reason for it is the re-arch is going to invalidate all current PRs, because it is going to introduce massive breaking changes to how AutoGPT works. Also not a good idea to beg for merge IMO.

Well, in the wiki it's also written that it can be a good idea to merge before the re-integration
https://github.com/Significant-Gravitas/Auto-GPT/wiki/Architecting#2-push-for-your-pr-to-be-merged-into-master-before-re-integration

But i understand that there are many changes wich are way more complex and critical than mine, and i'm perfectly ok to wait and eventually rewrite something if the mantainers require that.

Also... @ntindle asked for a test to coverage the new code.
I don't really know what can be a good way to make an unit test for this, since this is meant to connect to any external openai-compliant API.
It still uses all the core functions used for interacting with GPT3.5 and GPT4, is it really needed/useful?

**Comment by MarkSchmidty** [comment]
> Also not a good idea to beg for merge IMO.

I understand it's strange. But the linked Wiki article basically says to do exactly that. "[2. push for your PR to be merged into master before re-integration.](https://github.com/Significant-Gravitas/Auto-GPT/wiki/Architecting#2-push-for-your-pr-to-be-merged-into-master-before-re-integration)"

It's not my PR, but it does satisfy the Issues I've been advocating for since the early days of Auto-GPT. So I'm advocating for it to be merged before re-integration, per the linked wiki instructions.

**Comment by anonhostpi** [comment]
> Well, in the wiki it's also written that it can be a good idea to merge before the re-integration
> https://github.com/Significant-Gravitas/Auto-GPT/wiki/Architecting#2-push-for-your-pr-to-be-merged-into-master-before-re-integration

Well, yes but the re-arch is already hot underway. They started it last night

> But i understand that there are many changes wich are way more complex and critical than mine, and i'm perfectly ok to wait and eventually rewrite something if the mantainers require that.

Don't get too discouraged. The re-arch was only started 24 hours ago, and they have already reworked the Agent, Plugin, Workspace, and Command structures. See #3856 and the re-arch GH issues tag

I give it a week, maybe few weeks tops, and PRs will start getting pulled frequently again.

**Comment by Pwuts** [comment]
This is a useful addition, although this PR already has conflicts with the memory system rework that is underway, and I expect that the goal of this PR will also be achieved by abstraction of the embedding model which is part of the re-arch. Is there an acute need for this change that can't wait a week?

**Comment by MarkSchmidty** [comment]
> This is a useful addition, although this PR already has conflicts with the memory system rework that is underway, and I expect that the goal of this PR will also be achieved by abstraction of the embedding model which is part of the re-arch. Is there an acute need for this change that can't wait a week?

I'm not sure what qualifies as an acute need. The real purpose of this PR is to support arbitrary (non OpenAI) models, which is one of the most reacted and commented feature requests (if not the most) in the Issues. It seems important that any re-architecting takes this into consideration so as not to create unnecessary technical debt.

Note: The PR has two parts. 1. A configurable endpoint proxy URL and 2. Configurable embedding sizes. Both features are needed simultaneously to support arbitrary non-OpenAI models.

**Review by ntindle — DISMISSED** [review]
Pending Maintainer Disucssion

**Comment by DGdev91** [comment]
PR #4208 basically made all the work on embed_dim no more necessary.
Eventually, it could be a good idea to add a configuration for embed dim in the future memory modules/plugins, but it isn't in the scope of this PR anymore.

File: .env.template
Comment (Pwuts) [inline .env.template:25]: Please specify the URL format, e.g. with or without scheme or trailing slash

**Review by Pwuts — CHANGES_REQUESTED** [review]

**Comment by Pwuts** [comment]
Personal note: sorry it has been taking so long. We're swamped in a huge re-arch and a ton of PRs and we're still busy figuring out a scalable team workflow. Thanks for staying with us :)

File: .env.template
Comment (DGdev91) [reply .env.template:25]: Ok, done. i added an example

**Review by Pwuts — APPROVED** [review]

**Review by merwanehamadi — APPROVED** [review]

**Comment by DGdev91** [comment]
> Pending Maintainer Disucssion

@Pwuts just approved, if it's ok for you too it can be merged

**Comment by Pwuts** [comment]
@merwanehamadi I wouldn't know how this PR would change behavior.. Can you figure out why it's giving this comment?

**Comment by merwanehamadi** [comment]
@Pwuts if the PR changed the behaviour once, currently it will keep sending the behaviour label. So this PR changed the behaviour in the past, most likely
Options: 
- remove behaviour label change
- fix this label this is tricky to do. We need to identify whether the last cassette in the cassette file is being used

**Comment by Pwuts** [comment]
We're aware that this functionality was broken in #4803 and are working to fix it

**Pull request merged**
